### PR TITLE
8299125: UnifiedOopRef in JFR leakprofiler should treat thread local oops correctly

### PR DESCRIPTION
**Pull request by xmas92**
In a similar vain to [JDK-8299089](https://bugs.openjdk.org/browse/JDK-8299089) there are different semantics for OopStorage and thread local oops. UnifiedOopRef in the jfr leakprofiler must be able to distinguish these and treat them appropriately.

Testing: Running test at the moment. Has been running on the generational branch of ZGC for over three months. Not many tests exercise the leakprofiler. x86_32 has mostly been tested with GHA. 

Note: The accessBackend changes are only there to facilitate comparing OopLoadProxy objects directly with nullptr instead of creating and comparing with an `oop()` / `oop(NULL)`. Can be backed out of this PR and put in a different RFE instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299125](https://bugs.openjdk.org/browse/JDK-8299125): UnifiedOopRef in JFR leakprofiler should treat thread local oops correctly

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [e0d2eea4](https://git.openjdk.org/jdk/pull/11741/files/e0d2eea43a4d70760a0885ef838ec3e7fbc6aa69)
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11741/head:pull/11741` \
`$ git checkout pull/11741`

Update a local copy of the PR: \
`$ git checkout pull/11741` \
`$ git pull https://git.openjdk.org/jdk pull/11741/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11741`

View PR using the GUI difftool: \
`$ git pr show -t 11741`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11741.diff">https://git.openjdk.org/jdk/pull/11741.diff</a>

</details>

### CODE REVIEW

**Review by fisk — CHANGES_REQUESTED** [review]
Hmm. So we have IN_NATIVE accesses tagged as encode_in_native, and essentially AS_RAW accesses encoded as encode_non_barriered. I usually don't pick on naming, but the term "non-barriered" seems to stray away from already established terminology. I would prefer if the mentions of "non_barriered" changed to "as_raw" for consistency.

**Comment by xmas92** [comment]
> Hmm. So we have IN_NATIVE accesses tagged as encode_in_native, and essentially AS_RAW accesses encoded as encode_non_barriered. I usually don't pick on naming, but the term "non-barriered" seems to stray away from already established terminology. I would prefer if the mentions of "non_barriered" changed to "as_raw" for consistency.

I agree, changed the terms used.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by mgronlun** [comment]
This is introducing some new, very complicated tagging scheme. I don't understand it. What is 0b011 ?

File: src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ template <typename Delegate>
 void RootSetClosure<Delegate>::do_oop(oop* ref) {
   assert(ref != NULL, "invariant");
   assert(is_aligned(ref, HeapWordSize), "invariant");
-  if (*ref != NULL) {
+  if (NativeAccess<>::oop_load(ref) != nullptr) {
```
Comment (mgronlun) [inline src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp:51]: Why is this needed now?

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.hpp [hunk 2]
```diff
@@ -29,21 +29,33 @@
 #include "utilities/globalDefinitions.hpp"
 
 struct UnifiedOopRef {
+  static const uintptr_t tag_mask   = LP64_ONLY(0b111) NOT_LP64(0b011);
+  static const uintptr_t native_tag = 0b001;
+  static const uintptr_t raw_tag    = 0b010;
```
Comment (mgronlun) [inline src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.hpp:34]: What is "raw", compared to "native"?

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp [hunk 3]
```diff
@@ -42,47 +42,60 @@ inline T UnifiedOopRef::addr() const {
 // this specialization provides a workaround.
 template<>
 inline uintptr_t UnifiedOopRef::addr<uintptr_t>() const {
-  return _value & ~uintptr_t(3);
+  return (_value & ~tag_mask) LP64_ONLY(>> 1);
```
Comment (mgronlun) [inline src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: Why a shift now as well?

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp [hunk 4]
```diff
@@ -42,47 +42,60 @@ inline T UnifiedOopRef::addr() const {
...
-inline UnifiedOopRef UnifiedOopRef::encode_in_native(const narrowOop* ref) {
+template <typename T>
+inline UnifiedOopRef create_with_tag(T ref, uintptr_t tag) {
   assert(ref != NULL, "invariant");
-  UnifiedOopRef result = { reinterpret_cast<uintptr_t>(ref) | 3 };
-  assert(result.addr<narrowOop*>() == ref, "sanity");
+  assert(((reinterpret_cast<uintptr_t>(ref) LP64_ONLY(<< 1)) & UnifiedOopRef::tag_mask) == 0, "Unexpected low-order bits");
+  LP64_ONLY(assert(((reinterpret_cast<uintptr_t>(ref)) & (1ull << 63)) == 0, "Unexpected high-order bit"));
+  UnifiedOopRef result = { (reinterpret_cast<uintptr_t>(ref) LP64_ONLY(<< 1)) | tag };
+  assert(result.addr<T>() == ref, "sanity");
```
Comment (mgronlun) [inline src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:70]: This is crazy hard to read.

File: src/hotspot/share/oops/accessBackend.hpp [hunk 5]
```diff
@@ -1256,6 +1256,14 @@ namespace AccessInternal {
     inline bool operator !=(const T& other) const {
       return load<decorators | INTERNAL_VALUE_IS_OOP, P, T>(_addr) != other;
     }
+
+    inline bool operator ==(std::nullptr_t) const {
```
Comment (mgronlun) [inline src/hotspot/share/oops/accessBackend.hpp:1260]: why std::nullptr_t?

**Review by mgronlun — CHANGES_REQUESTED** [review]

File: src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp (hunk 1 above)
Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp:51]: The GC requires that load barriers to access oops in native storage. Currently the raw load works on all GCs as it only checks for null and all GCs encode null as NULL. But it is semantically wrong, and will not work with generational ZGC where there exist colored null which is not encoded as NULL.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.hpp (hunk 2 above)
Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.hpp:34]: Native is native storage (not in java heap) which requires GC barriers. Raw is raw storage which does not require GC barriers. This is thread local storage, such as the handle area, stack, etc.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 3 above)
Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: I do not remember exactly. But I think there was some ref which on 64 bit which had the third bit set. That is this assert failed:
`assert((reinterpret_cast<uintptr_t>(ref) & UnifiedOopRef::tag_mask) == 0, "Unexpected low-order bits");`
I have to remove the shift and test again. I think it is strange given that `ObjectAlignmentInBytes >= 8` but we added the shift for some reason. Or maybe I am just missing something obvious.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 4 above)
Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:70]: Yeah. Refactoring out `reinterpret_cast<uintptr_t>(ref)` into a variable might help a bit at least. The assert just want to check that the lowest two bits are not set (the tagging does not destroy information) and that on 64 bit platforms the highest bit is not set (the shift does not destroy information).

Not sure if there is a better way to express those two invariants.

**Comment by xmas92** [comment]
> This is introducing some new, very complicated tagging scheme. I don't understand it. What is 0b011 ?

We have three storage/access types in the vm, Native, Raw, Heap. The two lower bits encode this. The only valid low bits are Heap: 0b00, Native: 0b01 and Raw: 0b10. Then each storage area can be use compressed oops or not. Which is the third bit.

`0b011` would be an invalid tag.

File: src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp (hunk 1 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp:51]: Thanks. Should not the same construct be used in RawRootClosure? Or is it not needed there because the raw references to oops are not loaded through the Access Api?

Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp:51]: Raw storage has no GC barriers so a null oop there must be NULL. The RawAccess API is really only used when dealing with narrowOop or templated generic code which can be either. `RawAccess<>::oop_load(oop* ref)` will just become a `*ref`. I think I used RawAccess everywhere initially ,but @fisk thought that it should not be used for raw oop*, not sure if there were some other reason behind it than it being unnecessary / obfuscating.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 3 above)
Comment (fisk) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: ObjectAlignmentInBytes is 4 by default on 32 bit VMs.

File: src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp (hunk 1 above)
Comment (fisk) [reply src/hotspot/share/jfr/leakprofiler/chains/rootSetClosure.cpp:51]: I think what I might have said is that RawAccess helps with compressed oop conversion and memory ordering sprinklings. If you don't need to do anything special with either of those and you just want a plain dereference of an oop*, then just doing a plain dereference achieves the same thing with a lot less templates. Using RawAccess can make it look like there is more to it than there really is. It's perfectly correct, but maybe just not needed and makes you scratch your head a bit when really there is nothing special to see at all.

**Comment by mgronlun** [comment]
> > This is introducing some new, very complicated tagging scheme. I don't understand it. What is 0b011 ?
> 
> We have three storage/access types in the vm, Native, Raw, Heap. The two lower bits encode this. The only valid low bits are Heap: 0b00, Native: 0b01 and Raw: 0b10. Then each storage area can be use compressed oops or not. Which is the third bit.
> 
> `0b011` would be an invalid tag.

I see now it is the mask on 32-bit.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 3 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: @fisk Why is the shifts needed on 64-bits?

Comment (mgronlun) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: Maybe the shift is needed in order to handle unaligned oops, as for code blob oops? The existing root processing logic has skipped code blob oops, mainly because they were hard to encode. The rationale was that they are not roots in an of themselves. Now, we have seen several instances where chains are not found on iterations, so maybe the assumption on skipping code blob oops is wrong? With this encoding scheme in place, it would be possible to pass a closure for processing the code blob oops as well.

Comment (xmas92) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: Sorry for the tardy response. 

I think I forgot what the code was actually doing at the time of writing my previous answer. We are encoding ` narrowOop*` and `oop*` here not `narrowOop`/`oop`. And `narrowOop*` are 4 byte aligned.

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 4 above)
Comment (stefank) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:75]: Maybe it would be easier to read the code if it was rewritten as (uncompiled):
```
  uintptr_t value = reinterpret_cast<uintptr_t>(ref);

#ifdef _LP64
  // tag_mask is 3 bits. When ref is a narrowOop* we only have 2 alignment
  // bits, because of the 4 byte alignment of compressed oops addresses.
  // Shift up to make way for one more bit.
  assert(value & (1ull << 63)) == 0, "Unexpected high-order bit"));
  value <<= 1;
#endif

  UnifiedOopRef result = { value | tag };
  assert(result.addr<T>() == ref, "sanity");
```

File: src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp (hunk 3 above)
Comment (mgronlun) [reply src/hotspot/share/jfr/leakprofiler/utilities/unifiedOopRef.inline.hpp:45]: Ok.

**Review by mgronlun — APPROVED** [review]
